### PR TITLE
Azure App Service PAAS deployment

### PR DESCRIPTION
**Pull request by theprash**
Adds ARM template and Azure deployment to FAKE script for #51.

I needed to introduce an environment variable to allow changing the public content path because the Azure deployment I was using required putting the server at the root.

This depends on #64.

### CODE REVIEW

**Comment by isaacabraham** [comment]
Closes #51

File: Content/.template.config/template.json [hunk 1]
```diff
@@ -81,6 +81,12 @@
       "dataType": "bool",
       "defaultValue": "false",
       "description": "use NPM instead of default Yarn for JS package management"
+    },
+    "Azure": {
```
Comment (theimowski) [inline Content/.template.config/template.json]: How about introducing a `choice` parameter, say "Deploy" with two options for now: `Docker` and `Azure` - both options do not have much sense if combined

File: Content/arm-template.json [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        "appServicePlan": "[concat('safe-', parameters('environment'), '-web-host')]",
+        "web": "[concat('safe-', parameters('environment'), '-web')]"
+    },
+    "resources": [
+        {
+            "type": "Microsoft.Web/serverfarms",
+            "sku": { "name": "F1" },
+            "name": "[variables('appServicePlan')]",
+            "apiVersion": "2016-09-01",
+            "location": "West Europe"
```
Comment (theimowski) [inline Content/arm-template.json]: I have no experience with ARM but wondering whether all of the options here, like location should be hard-coded?

File: Content/src/Server/Server.fsproj [hunk 3]
```diff
@@ -1,5 +1,5 @@
 ﻿<?xml version="1.0" encoding="utf-8"?>
-<Project Sdk="Microsoft.NET.Sdk">
+<Project Sdk="Microsoft.NET.Sdk.Web">
```
Comment (theimowski) [inline Content/src/Server/Server.fsproj]: What's the difference?

File: Content/build.fsx [hunk 4]
```diff
@@ -113,8 +112,71 @@ Target "Docker" (fun _ ->
...
+  
+  let appName = deploymentOutputs.Value.WebAppName.value
+  let appPassword = deploymentOutputs.Value.WebAppPassword.value
+  let destinationUri = sprintf "https://%s.scm.azurewebsites.net/api/zipdeploy" appName
+  tracefn "Uploading %s to %s" zipFile destinationUri
+  let client = new Net.WebClient(Credentials = Net.NetworkCredential("$" + appName, appPassword))
+  client.UploadData(destinationUri, IO.File.ReadAllBytes zipFile) |> ignore
+)
+
+"Publish" ==> "DeployWebApp"
```
Comment (theimowski) [inline Content/build.fsx]: Can we put that into a single chain?

File: Content/build.fsx [hunk 5]
```diff
@@ -124,6 +186,11 @@ Target "Docker" (fun _ ->
   ==> "Docker"
 #endif
 
+#if (Azure)
+"Clean" ==> "Publish" ==> "DeployWebApp"
```
Comment (theimowski) [inline Content/build.fsx]: as above

**Comment by theimowski** [comment]
Started reviewing, but didn't yet manage to test myself

File: Content/.template.config/template.json (hunk 1 above)
Comment (isaacabraham) [reply Content/.template.config/template.json]: That makes sense to me as well.

File: Content/arm-template.json (hunk 2 above)
Comment (isaacabraham) [reply Content/arm-template.json]: The SKU and Location can both be parameterised (the Location definitely). @theprash The script is already using the `Region` type, which contains consts for all Azure regions, for the `deployment` object so you can use that as a means to supply this as an input parameter to the template. The SKU I don't think is available in code - [this template](https://github.com/Azure/azure-quickstart-templates/blob/master/201-web-app-github-deploy/azuredeploy.json) has an example of showing the different SKUs available and how to set them as input values.

File: Content/.template.config/template.json (hunk 1 above)
Comment (theprash) [reply Content/.template.config/template.json]: Agreed. I've made that change.

File: Content/.template.config/template.json [hunk 6]
```diff
@@ -125,11 +138,15 @@
           "condition": "(Fulma != \"login\")"
         },
         {
-          "exclude": "**/Dockerfile",
+          "exclude": "Dockerfile",
           "condition": "(!Docker)"
```
Comment (theimowski) [inline Content/.template.config/template.json]: the condition should now change to `"(Deploy  != \"docker\")"`

File: Content/.template.config/template.json [hunk 7]
```diff
@@ -125,11 +138,15 @@
...
         },
         {
-          "exclude": "**/Dockerfile",
+          "exclude": "Dockerfile",
           "condition": "(!Docker)"
         },
         {
-          "exclude": "**/yarn.lock",
+          "exclude": "arm-template.json",
+          "condition": "(!Azure)"
```
Comment (theimowski) [inline Content/.template.config/template.json]: similar to above

File: Content/src/Server/Server.fsproj (hunk 3 above)
Comment (theimowski) [reply Content/src/Server/Server.fsproj]: So that would be applicable only to `Azure` option?  Should this be a conditional then?

Comment (isaacabraham) [reply Content/src/Server/Server.fsproj]: I thought this would've been conditional, yes. @theprash thoughts?

File: Content/arm-template.json (hunk 2 above)
Comment (theprash) [reply Content/arm-template.json]: I've added a pricingTier parameter for the SKU.

**Comment by theimowski** [comment]
Is the `public_path` env var a must?

File: Content/build.fsx (hunk 4 above)
Comment (theprash) [reply Content/build.fsx]: Technically this is what the dependencies are (the ARM deploy only sets up Azure infrastructure) but @isaacabraham has already put them in a single chain and I'm happy with it.

File: Content/.template.config/template.json (hunk 6 above)
Comment (theprash) [reply Content/.template.config/template.json]: Whoops I missed those. Fix now, thanks.

File: Content/build.fsx [hunk 8]
```diff
@@ -106,13 +112,77 @@ Target "Docker" (fun _ ->
...
+let environment = getBuildParamOrDefault "environment" (Guid.NewGuid().ToString().ToLower().Split '-' |> Array.head)
+let location = getBuildParamOrDefault "location" Region.EuropeWest.Name
+
+let mutable deploymentOutputs : ArmOutput option = None
+
+Target "ArmTemplate" (fun _ ->
+  let armTemplate = @"arm-template.json"
+  let resourceGroupName = "safe-" + environment
+  let subscriptionId = try getBuildParam "subscriptionId" |> Guid.Parse with _ -> failwith "Invalid Subscription ID. This should be your Azure Subscription ID."
+  let clientId = try getBuildParam "clientId" |> Guid.Parse with _ -> failwith "Invalid Client ID. This should be the Client ID of a Native application registered in Azure with permission to create resources in your subscription."
```
[the remark this answers is not in the document]
Comment (isaacabraham) [reply Content/build.fsx]: This needs to be documented properly but you need to create an app in Active Directory, set it to "Native" and then give it permissions to Windows Azure. The application id is the client id you need to use.

**Comment by isaacabraham** [comment]
@theimowski what should we use instead? Environment variables usually work quite well across most systems I imagine.

**Comment by theprash** [comment]
@theimowski, We needed to be able to configure the public folder path in order for the Azure deploy to work, at least with the method that we're using here. The server had to be at the root of the published folder so it could not refer to a "Client" directory outside of itself. I discussed with @isaacabraham a few different ways we could do this including adding a config file, which is not well supported yet by FAKE for .NET Core. We thought an environment variable would be the least intrusive method for now. It can be set easily for the app in Azure and other systems as Isaac mentioned above.

**Comment by theimowski** [comment]
Sorry I didn't make myself clear. I meant to ask if its required to have any variance for the `public_path` and if so why's that?

**Comment by theimowski** [comment]
Ok thanks, then again I'd it in for making that conditional just for azure option

**Comment by isaacabraham** [comment]
Yeah. We could've deployed both the "client" and "server" folders as now but that makes things somewhat more complicated in terms of build / packaging I think. Perhaps we'll revisit this in the future but for now I think what is here should work (apart from your comment about making the environment bit Azure-only - I think that we can do that.)

File: Content/src/Server/Server.fsproj (hunk 3 above)
Comment (theprash) [reply Content/src/Server/Server.fsproj]: @theimowski @isaacabraham, I thought it might be better to avoid a conditional branch on the project type just to reduce complexity a little bit if there's little or no downside in using the `.Web` version for all versions of the template. Aside from the web.config file, it also adds some extra entries in `Server.deps.json` which seem to not be very important, but I don't know what they're for.

I'm happy to make it conditional though...?

**Comment by isaacabraham** [comment]
@theprash the currently docker file shows up when you do a `-D azure`.

**Comment by isaacabraham** [comment]
There's now a "fix" in this branch that means that default files show up correctly. We can remove it once the PR for Saturn is accepted.

File: Content/src/Server/Server.fsproj (hunk 3 above)
Comment (theprash) [reply Content/src/Server/Server.fsproj]: OK I've done it 😀

**Comment by theprash** [comment]
I believe everything raised here has now been resolved.

**Comment by isaacabraham** [comment]
@theimowski I've now created a (lengthy) guide on the docs site in the template-deployment branch. Can you give it a run through and see if it works / makes sense for you? We can then release this alongside that.

**Comment by isaacabraham** [comment]
I've now also added support for Suave, which is hosted in App Service slightly differently than Giraffe.

**Comment by theimowski** [comment]
I'll have a look on Tuesday

File: Content/.template.config/template.json [hunk 9]
```diff
@@ -150,7 +181,8 @@
           }
         }
       ]
-    },"**/*.fsx": {
+    },
+    "**/*.fs*": {
```
Comment (theimowski) [inline Content/.template.config/template.json]: is this change intentional?

**Comment by theimowski** [comment]
I started testing the `appservice` target and run into some dll hell issue:

```
Finished Target: Bundle
Starting Target: ArmTemplate (==> Bundle)
Deploying template 'arm-template.json' to resource group 'safe-795cba0b' in subscription '{subscriptionId}'...
Redirect assembly from 'System.Net.Http, Version=4.1.1.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' to 'System.Net.Http, Version=4.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a'
To sign in, use a web browser to open the page https://microsoft.com/devicelogin and enter the code {code} to authenticate.
Running build failed.
Error:
System.IO.FileNotFoundException: Could not load file or assembly 'Microsoft.Rest.ClientRuntime.Azure, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35' or one of its dependencies. The system cannot find the file specified.
File name: 'Microsoft.Rest.ClientRuntime.Azure, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35'
   at Microsoft.Azure.Management.ResourceManager.Fluent.ResourceManager.Authenticated..ctor(RestClient restClient)
   at FSI_0005.Cit.Helpers.Arm.authenticateDevice@128-2.Invoke(AuthenticationResult _arg2)
   at Microsoft.FSharp.Control.AsyncBuilderImpl.args@825-1.Invoke(a a)
--- End of stack trace from previous location where exception was thrown ---
   at Microsoft.FSharp.Control.AsyncBuilderImpl.commit[a](AsyncImplResult`1 res)
   at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronouslyInCurrentThread[a](CancellationToken token, FSharpAsync`1 computation)
   at Microsoft.FSharp.Control.CancellationTokenOps.RunSynchronously[a](CancellationToken token, FSharpAsync`1 computation, FSharpOption`1 timeout)
   at Microsoft.FSharp.Control.FSharpAsync.RunSynchronously[T](FSharpAsync`1 computation, FSharpOption`1 timeout, FSharpOption`1 cancellationToken)
   at FSI_0005.Build.clo@89-12.Invoke(Unit _arg1)
   at Fake.TargetHelper.runSingleTarget(TargetTemplate`1 target) in D:\code\fake\src\app\FakeLib\TargetHelper.fs:line 626

Assembly manager loaded from:  C:\Windows\Microsoft.NET\Framework\v4.0.30319\clr.dll
Running under executable  %home%\sandbox\test\packages\build\FAKE\tools\FAKE.exe
--- A detailed error log follows.

=== Pre-bind state information ===
LOG: DisplayName = Microsoft.Rest.ClientRuntime.Azure, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
 (Fully-specified)
LOG: Appbase = file:///%home%/sandbox/test/packages/build/FAKE/tools/
LOG: Initial PrivatePath = NULL
Calling assembly : Microsoft.Azure.Management.ResourceManager.Fluent, Version=1.0.0.60, Culture=neutral, PublicKeyToken=31bf3856ad364e35.
===
LOG: This bind starts in LoadFrom load context.
WRN: Native image will not be probed in LoadFrom context. Native image will only be probed in default load context, like with Assembly.Load().
LOG: Using application configuration file: %home%\sandbox\test\packages\build\FAKE\tools\FAKE.exe.Config
LOG: Using host configuration file:
LOG: Using machine configuration file from C:\Windows\Microsoft.NET\Framework\v4.0.30319\config\machine.config.
LOG: Post-policy reference: Microsoft.Rest.ClientRuntime.Azure, Version=3.0.0.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/FAKE/tools/Microsoft.Rest.ClientRuntime.Azure.DLL.
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/FAKE/tools/Microsoft.Rest.ClientRuntime.Azure/Microsoft.Rest.ClientRuntime.Azure.DLL.
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/FAKE/tools/Microsoft.Rest.ClientRuntime.Azure.EXE.
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/FAKE/tools/Microsoft.Rest.ClientRuntime.Azure/Microsoft.Rest.ClientRuntime.Azure.EXE.
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/Microsoft.Azure.Management.ResourceManager.Fluent/lib/netstandard1.3/Microsoft.Rest.ClientRuntime.Azure.DLL.
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/Microsoft.Azure.Management.ResourceManager.Fluent/lib/netstandard1.3/Microsoft.Rest.ClientRuntime.Azure/Microsoft.Rest.ClientRuntime.Azure.DLL.
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/Microsoft.Azure.Management.ResourceManager.Fluent/lib/netstandard1.3/Microsoft.Rest.ClientRuntime.Azure.EXE.
LOG: Attempting download of new URL file:///%home%/sandbox/test/packages/build/Microsoft.Azure.Management.ResourceManager.Fluent/lib/netstandard1.3/Microsoft.Rest.ClientRuntime.Azure/Microsoft.Rest.ClientRuntime.Azure.EXE.


---------------------------------------------------------------------
Build Time Report
---------------------------------------------------------------------
Target              Duration
------              --------
Clean               00:00:00.1522493
InstallDotNetCore   00:00:00.1603336
InstallClient       00:00:02.2953899
Build               00:00:20.5715311
Bundle              00:00:03.3486279
ArmTemplate         Failure
Total:              00:01:24.0550037
---------------------------------------------------------------------
Status:             Failure
---------------------------------------------------------------------
---------------------------------------------------------------------
```

Not the binding redirect for System.Net.Http at the beginning - not sure if that's desired?

File: Content/.template.config/template.json (hunk 9 above)
Comment (theimowski) [reply Content/.template.config/template.json]: I can see you also now use `//#if` instead of `#if` in `.fsx` script - is that this change that allowed that? Maybe we can get rid of this entry completely then?

**Comment by isaacabraham** [comment]
@theimowski The binding redirect with System.Net.Http is a red herring here I think. I've seen this issue this weekend with Microsoft.Rest.ClientRuntime.Azure. The "fix" is (and don't laugh) to make a change to the FAKE script e.g. add a new line anywhere in the script, and then re-run it. It then works.

I have no idea why this is happening or how to prevent it from occurring in the first place.

File: Content/.template.config/template.json (hunk 9 above)
Comment (isaacabraham) [reply Content/.template.config/template.json]: @theprash one for you.

**Comment by theimowski** [comment]
WTF, do we know why's that? If not the "fix" should be documented for sure

**Comment by isaacabraham** [comment]
I'm also considering moving away from the SDK and shelling out to the `az` command line tool instead (and we would install that during the fake script).

**Comment by theimowski** [comment]
Yes it worked :) Something FAKE-related? @matthid any ideas why resaving a fake script may have impact on assembly binding? This is using FAKE 4

**Comment by matthid** [comment]
So does this happen again after that and you need to make another change or does it always work now?

**Comment by theimowski** [comment]
Works now without need to make more changes. The `Redirect assembly from 'System.Net.Http` warning is still there, but works anyway

File: Content/.template.config/template.json (hunk 9 above)
Comment (theimowski) [reply Content/.template.config/template.json]: Right, makes sense. So you commented #if in fsx for better editor support? I think I like it this way now

**Comment by matthid** [comment]
No idea, so it is not reproduce-able at all?

**Comment by theimowski** [comment]
LGTM - I managed to deploy Saturn and Suave versions to azure.
Awesome work, huge props @theprash and @isaacabraham !

Can we add a note to docs about the FAKE modify-file "fix" until we know what's the real issue here?

**Comment by matthid** [comment]
> Can we add a note to docs about the FAKE modify-file "fix" until we know what's the real issue here?

Or upgrade to latest version of fake ;)

**Comment by theimowski** [comment]
Yes that's probably a good idea anyway - added https://github.com/SAFE-Stack/SAFE-template/issues/70

**Comment by theimowski** [comment]
This is now released in 0.15
Thanks again

**Comment by jarlestabell** [comment]
Amazing work, thanks! :)
BTW: I also got the ```Could not load file or assembly 'Microsoft.Rest.ClientRuntime.Azure, Version=3.0.0.0...``` error at first and applied the fix (dummy edit of the Fake script). And then it worked.

**Comment by isaacabraham** [comment]
@jarlestabell yeah. It's annoying - hopefully moving to FAKE 5 will "fix" this, but we don't know exactly why it's happening :-/
